### DISCUSSION TITLE
How to make a toast prompt pops up on the android app by using briefcase?

### DISCUSSION THREAD

**Discussion by ccinsz** [beeware/briefcase#1497] [Q&A]
I want to make a toast prompt pops up on the mobile phone system. How I need to do?

**Comment by mhsmith** [comment]
You can do it as shown here: https://github.com/chaquo/chaquopy/blob/da72f66abdb1a375887ff84f93be7efee6ae76a8/demo/app/src/main/python/chaquopy/demo/ui_demo.py#L80

Instead of `activity` you can use `self._impl.native`, where `self` is your App object. And instead of the `R.string` reference, you can simply pass a string.

**Reply by ccinsz** [reply]
This toast only display on its own app. Is there any toast that can display on the system?

**Reply by ccinsz** [reply]
Or how to call the system's floating box to notify some information?

**Reply by mhsmith** [reply]
On Android the word "toast" is only used to refer to the temporary non-interactive popups that usually appear at the bottom of the screen. But if you're talking about notifications in the top bar, there used to be an example of that in the same file, but it was removed in [this commit](https://github.com/chaquo/chaquopy/commit/c6a496acc8522bc9621ce6a5fee2230145b4213c) because Android now requires apps to ask for permission before showing notifications.

Unfortunately I don't have any example code showing how to ask for permissions from Python. You may be able to adapt a Java/Kotlin example using the `intent_result` method described in https://github.com/beeware/toga/issues/1798. If you make any progress on that, please post it on #547.

**Reply by ccinsz** [reply]
What should R.drawable.ic_launcher be changed to? In the android it can't know R.

**Reply by mhsmith** [reply]
You can import it from the package named after the [`bundle` in your pyproject.toml file](https://briefcase.readthedocs.io/en/latest/reference/configuration.html#bundle). For example:
```
from com.example.myapp import R
```

**Reply by ccinsz** [reply]
I made a demo for Android 8 or above. Add `<uses-permission android:name="android.permission.POST_NOTIFICATIONS" />` in the AndroidManifest.xml. Click the button to enter the permission settings page to set notification permissions. After setting the permission notification, click the button again and a notification will pop up. 

Does it need to be posted post on https://github.com/beeware/briefcase/issues/547 ?

**Reply by freakboy3742** [reply]
The Briefcase ticket you've referenced is a feature request for adding a Briefcase-managed mechanism for the permissions that you've added manually to AndroidManifest.xml. That feature doesn't exist yet; there's no reason to "post" anything unless you've got an implementation of that feature, or you have a particularly unusual usage that you think the ticket description doesn't cover.

**Reply by mhsmith** [reply]
It may not be directly relevant to the Briefcase part of the mechanism, but please post it anyway, as it may be useful to other people.

**Reply by ccinsz** [reply]
Ok, post it here.

```
import toga
from toga.style import Pack
from toga.style.pack import COLUMN
from android.content import Context
from com.example.demo import R
from androidx.core.app import NotificationCompat
from android.os import Build
from android.app import NotificationChannel
from android.app import NotificationManager
from android.content import Intent
from android.provider import Settings


class demo(toga.App):

    def startup(self):
        self.main_window = toga.MainWindow(title=self.formal_name)
        btn = toga.Button("Click me", on_press=self.my_callback)
        main_box = toga.Box(style=Pack(flex=1, direction=COLUMN))
        main_box.add(btn)

        self.main_window.content = main_box
        self.main_window.show()

    def my_callback(self, widget):
        manager = self._impl.native.getSystemService(Context.NOTIFICATION_SERVICE)
        if Build.VERSION.SDK_INT >= Build.VERSION_CODES.O:
            channel = NotificationChannel("default", "Default", NotificationManager.IMPORTANCE_HIGH)
            manager.createNotificationChannel(channel)

        if not manager.areNotificationsEnabled():
            intent = Intent(Settings.ACTION_APP_NOTIFICATION_SETTINGS)
            intent.putExtra(Settings.EXTRA_APP_PACKAGE, self._impl.native.getPackageName())
            self._impl.native.startActivity(intent)

        builder = NotificationCompat.Builder(self._impl.native, "default")
        builder.setSmallIcon(R.drawable.launch)
        builder.setContentTitle("mytest")
        builder.setContentText("test")
        manager.notify(1, builder.getNotification())


def main():
    return demo()
```

**Reply by mhsmith** [reply]
Thanks.
